### DISCUSSION TITLE
Running ClassGraph within custom rule in Maven Enforcer plugin doesn't find target/classes ?

### DISCUSSION THREAD

**Discussion by lejtemxviw** [classgraph/classgraph#752] [Q&A]
Hi!  I am trying to write a custom rule for the Maven Enforcer plugin that uses ClassGraph to find classes of interest, then runs some validation logic on them to approve/fail the build.  However, when I run this rule as part of a Maven build, ClassGraph outputs info listing all the jar files that are dependencies on the project - but doesn't include target/classes, which is where Maven just put all the compiled code.  If I modify the ClassGraph construction to use `.overrideClasspath("target/classes")` then it does find all the classes in the current build - which is what I want, but it doesn include anything other than that folder, which I don't think I want to then ignore :-(.
I did not find a way to just add `"target/classes"` to the list of classpath locations that ClassGraph will read - am I missing something here?
Thank you!

**Comment by lukehutch** [comment]
Honestly I loathe Maven, it has caused me nothing but pain! But if I had to guess, you need to run the enforcer plugin in a later phase of the build.

**Reply by lejtemxviw** [reply]
Ah!  Sorry - I had thought that, too - so my first attempt at a fix was to move the execution of my rule to the test phase, when I was sure the main code would have already been compiled ... But it still failed the same way.  Using the `overrideClasspath()` (still during the test phase) does work.  The documentation says ClassGraph is smart enough to detect automatically a "classes/" folder to look inside of - but it looks like it doesn't know to look in "target/classes/" too?
In retrospect, maybe it makes sense that Maven's own class loader hierarchy should not include "target/classes", since that's the code it's compiling - and maybe that affects ClassGraph, since it's running as part of Maven when used in an Enforcer custom rule?
But I still would like it to see "target/classes" _plus_ all the jars in the Maven class loader hierarchy...  (Or, my project's class loader hierarchy, if those two are different...)

**Comment by lukehutch** [comment]
Correct, `target/classes` should be in the classpath. You should print out the classpath (without overriding it), or turn on verbose logs, to figure out why this directory is not present or not being scanned.

**Reply by lejtemxviw** [reply]
I did a little research - according to this page on Maven classloading, Maven plugins do not have access to the build directory, or project dependencies: https://maven.apache.org/guides/mini/guide-maven-classloading.html (towards the end of the page).
However, Maven does provide a mechanism that allows the plugin to ask what the project's compile build path would be.
I am thinking that, since my custom rule is running within the Enforcer plugin, it qualifies as a plugin - so I will try querying the compile classpath to initialize my ClassGraph scan...
